rclcpp/include/rclcpp/experimental/executors/events_executor/priority_events_queue.hpp

```diff
@@ -0,0 +1,169 @@
+// Copyright 2023 iRobot Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__EXPERIMENTAL__EXECUTORS__EVENTS_EXECUTOR__PRIORITY_EVENTS_QUEUE_HPP_
+#define RCLCPP__EXPERIMENTAL__EXECUTORS__EVENTS_EXECUTOR__PRIORITY_EVENTS_QUEUE_HPP_
+
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <queue>
+#include <utility>
+#include <vector>
+
+#include "rclcpp/experimental/executors/events_executor/events_queue.hpp"
+
+namespace rclcpp
+{
+namespace experimental
+{
+namespace executors
+{
+
+struct PriorityEvent
+{
+  int priority;
+  rclcpp::experimental::executors::ExecutorEvent event;
+};
+
+struct ComparePriorities : public std::binary_function<PriorityEvent, PriorityEvent, bool>
+{
+  _GLIBCXX14_CONSTEXPR
+  bool
+  operator()(const PriorityEvent & __x, const PriorityEvent & __y) const
+  {return __x.priority > __y.priority;}
+};
+
+/**
+ * @brief This class implements an EventsQueue as a simple wrapper around a std::priority_queue.
+ * It does not perform any checks about the size of queue, which can grow
+ * unbounded without being pruned.
+ */
+class PriorityEventsQueue : public EventsQueue
+{
+public:
+  RCLCPP_PUBLIC
+  PriorityEventsQueue()
+  {
+    // Default callback to extract priority from event
+    extract_priority_ = [](const rclcpp::experimental::executors::ExecutorEvent & event) {
+        (void)(event);
+        return 0;
+      };
+  }
+
+  RCLCPP_PUBLIC
+  explicit PriorityEventsQueue(
+    std::function<int(const rclcpp::experimental::executors::ExecutorEvent &)> extract_priority)
+  : extract_priority_(extract_priority) {}
```

Comment:
here we should check whether the passed `extract_priority` function is callable, and throw an exception if it's not